File: src/cli/output.rs
//! Terminal output formatting

use console::{style, Term};
use indicatif::{ProgressBar, ProgressStyle};

/// Output formatter for CLI
pub struct Output {
    term: Term,
    quiet: bool,
    verbose: bool,
}

impl Output {
    /// Create a new output formatter
    pub fn new(quiet: bool, verbose: bool) -> Self {
        Self {
            term: Term::stderr(),
            quiet,
            verbose,
        }
    }

    /// Print an info message
    pub fn info(&self, message: &str) {
        if !self.quiet {
            let _ = self
                .term
                .write_line(&format!("{} {}", style("==>").cyan().bold(), message));
        }
    }

    /// Print a success message
    pub fn success(&self, message: &str) {
        if !self.quiet {
            let _ = self
                .term
                .write_line(&format!("{} {}", style("✓").green().bold(), message));
        }
    }

    /// Print a warning message
    pub fn warning(&self, message: &str) {
        let _ = self.term.write_line(&format!(
            "{} {}",
            style("Warning:").yellow().bold(),
            message
        ));
    }

    /// Print an error message
    pub fn error(&self, message: &str) {
        let _ = self
            .term
            .write_line(&format!("{} {}", style("Error:").red().bold(), message));
    }

    /// Print a verbose/debug message
    pub fn debug(&self, message: &str) {
        if self.verbose && !self.quiet {
            let _ = self
                .term
                .write_line(&format!("{} {}", style("DEBUG:").dim(), message));
        }
    }

    /// Print a plain message
    pub fn print(&self, message: &str) {
        if !self.quiet {
            println!("{}", message);
        }
    }

    /// Print a styled package name
    pub fn package_name(name: &str) -> String {
        style(name).green().bold().to_string()
    }

    /// Print a styled version
    pub fn version(version: &str) -> String {
        style(version).cyan().to_string()
    }

    /// Create a progress bar for downloads
    pub fn download_progress(&self, total: u64, message: &str) -> ProgressBar {
        let pb = ProgressBar::new(total);
        pb.set_style(
            ProgressStyle::default_bar()
                .template("{msg}\n{spinner:.green} [{elapsed_precise}] [{bar:40.cyan/blue}] {bytes}/{total_bytes} ({eta})")
                .unwrap()
                .progress_chars("#>-"),
        );
        pb.set_message(message.to_string());
        pb
    }

    /// Create a spinner for indeterminate progress
    pub fn spinner(&self, message: &str) -> ProgressBar {
        let pb = ProgressBar::new_spinner();
        pb.set_style(
            ProgressStyle::default_spinner()
                .template("{spinner:.green} {msg}")
                .unwrap(),
        );
        pb.set_message(message.to_string());
        pb.enable_steady_tick(std::time::Duration::from_millis(100));
        pb
    }

    /// Print a table header
    pub fn table_header(&self, columns: &[(&str, usize)]) {
        if self.quiet {
            return;
        }

        let header: String = columns
            .iter()
            .map(|(name, width)| format!("{:width$}", name, width = *width))
            .collect::<Vec<_>>()
            .join("  ");

        println!("{}", style(&header).bold());
        println!("{}", style("-".repeat(header.len())).dim());
    }

    /// Print a table row
    pub fn table_row(&self, values: &[(&str, usize)]) {
        if self.quiet {
            return;
        }

        let row: String = values
            .iter()
            .map(|(value, width)| format!("{:width$}", value, width = *width))
            .collect::<Vec<_>>()
            .join("  ");

        println!("{}", row);
    }

    /// Print package info in a formatted way
    pub fn package_info(&self, name: &str, version: &str, desc: Option<&str>) {
        if self.quiet {
            return;
        }

        println!("{} {}", style(name).green().bold(), style(version).cyan());

        if let Some(desc) = desc {
            println!("  {}", desc);
        }
    }

    /// Print a list item
    pub fn list_item(&self, item: &str, details: Option<&str>) {
        if self.quiet {
            return;
        }

        print!("  {} {}", style("•").dim(), item);
        if let Some(details) = details {
            print!(" {}", style(details).dim());
        }
        println!();
    }

    /// Print a section header
    pub fn section(&self, title: &str) {
        if !self.quiet {
            println!();
            println!("{}", style(title).bold().underlined());
        }
    }

    /// Print caveats
    pub fn caveats(&self, caveats: &str) {
        if !self.quiet {
            println!();
            println!("{}", style("==> Caveats").yellow().bold());
            for line in caveats.lines() {
                println!("  {}", line);
            }
        }
    }

    /// Print a hint/suggestion
    pub fn hint(&self, message: &str) {
        if !self.quiet {
            println!("{} {}", style("Hint:").blue(), message);
        }
    }
}

impl Default for Output {
    fn default() -> Self {
        Self::new(false, false)
    }
}

/// Format a duration as human-readable
pub fn format_duration(secs: u64) -> String {
    if secs < 60 {
        format!("{}s", secs)
    } else if secs < 3600 {
        format!("{}m {}s", secs / 60, secs % 60)
    } else {
        format!("{}h {}m", secs / 3600, (secs % 3600) / 60)
    }
}

/// Format bytes as human-readable
pub fn format_bytes(bytes: u64) -> String {
    crate::storage::cache::format_bytes(bytes)
}

File: swift/Sources/ColdbrewKit/CLI/Output.swift
import Foundation

public struct Output {
    public typealias Writer = (String) -> Void

    private let quiet: Bool
    private let verbose: Bool
    private let stdout: Writer
    private let stderr: Writer

    public init(
        quiet: Bool = false,
        verbose: Bool = false,
        stdout: @escaping Writer = { Swift.print($0) },
        stderr: @escaping Writer = { message in
            FileHandle.standardError.write(Data((message + "\n").utf8))
        }
    ) {
        self.quiet = quiet
        self.verbose = verbose
        self.stdout = stdout
        self.stderr = stderr
    }

    public func info(_ message: String) {
        guard !quiet else { return }
        stderr("==> \(message)")
    }

    public func success(_ message: String) {
        guard !quiet else { return }
        stderr("✓ \(message)")
    }

    public func warning(_ message: String) {
        stderr("Warning: \(message)")
    }

    public func error(_ message: String) {
        stderr("Error: \(message)")
    }

    public func debug(_ message: String) {
        guard verbose, !quiet else { return }
        stderr("DEBUG: \(message)")
    }

    public func print(_ message: String) {
        guard !quiet else { return }
        stdout(message)
    }

    public static func packageName(_ name: String) -> String {
        name
    }

    public static func version(_ version: String) -> String {
        version
    }

    public func tableHeader(_ columns: [(String, Int)]) {
        guard !quiet else { return }
        let header = tableLine(columns)
        stdout(header)
        stdout(String(repeating: "-", count: header.count))
    }

    public func tableRow(_ values: [(String, Int)]) {
        guard !quiet else { return }
        stdout(tableLine(values))
    }

    public func packageInfo(name: String, version: String, description: String?) {
        guard !quiet else { return }
        stdout("\(Self.packageName(name)) \(Self.version(version))")
        if let description {
            stdout("  \(description)")
        }
    }

    public func listItem(_ item: String, details: String? = nil) {
        guard !quiet else { return }
        stdout("  • \(item)" + details.map { " \($0)" }.orEmpty)
    }

    public func section(_ title: String) {
        guard !quiet else { return }
        stdout("")
        stdout(title)
    }

    public func caveats(_ caveats: String) {
        guard !quiet else { return }
        stdout("")
        stdout("==> Caveats")
        for line in caveats.split(separator: "\n", omittingEmptySubsequences: false) {
            stdout("  \(line)")
        }
    }

    public func hint(_ message: String) {
        guard !quiet else { return }
        stdout("Hint: \(message)")
    }

    private func tableLine(_ values: [(String, Int)]) -> String {
        values.map { value, width in
            value.padding(toLength: max(value.count, width), withPad: " ", startingAt: 0)
        }.joined(separator: "  ")
    }
}

public func formatDuration(_ seconds: UInt64) -> String {
    if seconds < 60 {
        return "\(seconds)s"
    } else if seconds < 3_600 {
        return "\(seconds / 60)m \(seconds % 60)s"
    } else {
        return "\(seconds / 3_600)h \((seconds % 3_600) / 60)m"
    }
}

public func formatBytes(_ bytes: UInt64) -> String {
    let kb: UInt64 = 1_024
    let mb = kb * 1_024
    let gb = mb * 1_024

    if bytes >= gb {
        return String(format: "%.2f GB", Double(bytes) / Double(gb))
    } else if bytes >= mb {
        return String(format: "%.2f MB", Double(bytes) / Double(mb))
    } else if bytes >= kb {
        return String(format: "%.2f KB", Double(bytes) / Double(kb))
    } else {
        return "\(bytes) bytes"
    }
}

private extension Optional where Wrapped == String {
    var orEmpty: String {
        self ?? ""
    }
}
